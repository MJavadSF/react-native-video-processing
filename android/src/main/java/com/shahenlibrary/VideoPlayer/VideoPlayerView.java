package com.shahenlibrary.VideoPlayer;

import android.graphics.Bitmap;
import android.graphics.Matrix;
import android.media.MediaMetadataRetriever;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Handler;
import android.util.Base64;
import android.util.Log;
import android.widget.MediaController;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.events.Event;
import com.facebook.react.uimanager.events.RCTEventEmitter;
import com.yqritc.scalablevideoview.ScalableVideoView;
import com.yqritc.scalablevideoview.ScaleManager;
import com.yqritc.scalablevideoview.Size;
import com.shahenlibrary.Events.Events;
import com.shahenlibrary.Events.EventsEnum;

import java.io.ByteArrayOutputStream;
import java.io.IOException;

public class VideoPlayerView extends ScalableVideoView implements
        MediaPlayer.OnPreparedListener, MediaPlayer.OnErrorListener, MediaPlayer.OnBufferingUpdateListener,
        MediaPlayer.OnCompletionListener, MediaPlayer.OnInfoListener, LifecycleEventListener, MediaController.MediaPlayerControl {

    private ThemedReactContext themedReactContext;
    private RCTEventEmitter eventEmitter;
    private String mediaSource;
    private boolean playerPlaying = true;
    private String LOG_TAG = "RNVideoProcessing";
    private Runnable progressRunnable = null;
    private Handler progressUpdateHandler = new Handler();
    private MediaMetadataRetriever metadataRetriever = new MediaMetadataRetriever();
    private int progressUpdateHandlerDelay = 1000;
    private int videoStartAt = 0;
    private int videoEndAt = -1;


    public VideoPlayerView(ThemedReactContext ctx) {
        super(ctx);
        themedReactContext = ctx;
        eventEmitter = themedReactContext.getJSModule(RCTEventEmitter.class);
        themedReactContext.addLifecycleEventListener(this);
        setSurfaceTextureListener(this);
        initPlayerIfNeeded();

        progressRunnable = new Runnable() {
            @Override
            public void run() {
                if (mMediaPlayer != null && mMediaPlayer.isPlaying()) {
                    Log.d(LOG_TAG, "run: Send event");
                    WritableMap event = Arguments.createMap();
                    event.putDouble(Events.CURRENT_TIME, mMediaPlayer.getCurrentPosition() / 1000.0);
                    if (mMediaPlayer.getCurrentPosition() >= videoEndAt && videoEndAt != -1) {
                        Log.d(LOG_TAG, "run: End time reached");
                        mMediaPlayer.seekTo(videoStartAt);
                        if (!mMediaPlayer.isLooping()) {
                            Log.d(LOG_TAG, "run: Set loop");
                            pause();
                        }
                    }
                    eventEmitter.receiveEvent(getId(), EventsEnum.EVENT_PROGRESS.toString(), event);
                }

                progressUpdateHandler.postDelayed(progressRunnable, progressUpdateHandlerDelay);
            }
        };

        progressUpdateHandler.post(progressRunnable);
    }

    private void initPlayerIfNeeded() {
        if (mMediaPlayer != null) {
            return;
        }
        Log.d(LOG_TAG, "initPlayerIfNeeded");
        mMediaPlayer = new MediaPlayer();
        mMediaPlayer.setScreenOnWhilePlaying(true);
        mMediaPlayer.setOnVideoSizeChangedListener(this);
        mMediaPlayer.setOnErrorListener(this);
        mMediaPlayer.setOnPreparedListener(this);
        mMediaPlayer.setOnBufferingUpdateListener(this);
        mMediaPlayer.setOnCompletionListener(this);
        mMediaPlayer.setOnInfoListener(this);
    }

    public void setSource(final String uriString) {
        if (mediaSource != null && mediaSource.equals(uriString)) {
            return;
        }
        if (mMediaPlayer == null) {
            Log.d(LOG_TAG, "setSource: Media player is null");
            return;
        }
        reset();

        mediaSource = uriString;
        Log.d(LOG_TAG, "set source: " + mediaSource);

        initPlayerIfNeeded();

        try {
            if (uriString.startsWith("content://")) {
                Uri parsedUri = Uri.parse(mediaSource);
                setDataSource(themedReactContext, parsedUri);
                metadataRetriever.setDataSource(themedReactContext, parsedUri);
            } else {
                setDataSource(mediaSource);
                metadataRetriever.setDataSource(mediaSource);
            }
            prepare(this);

            if (playerPlaying && !mMediaPlayer.isPlaying()) {
                Log.d(LOG_TAG, "setSource: Start video at once");
                start();
            }

        } catch (IOException e) {
            e.printStackTrace();
            Log.d(LOG_TAG, "setSrc: ERROR");
        }
    }

    public void setPlay(final boolean shouldPlay) {
        Log.d(LOG_TAG, "setPlay: " + shouldPlay);
        playerPlaying = shouldPlay;

        if (mMediaPlayer == null) {
            Log.d(LOG_TAG, "setPlay: Player reference is null");
            return;
        }
        if (shouldPlay && !mMediaPlayer.isPlaying()) {
            start();
            Log.d(LOG_TAG, "setPlay: START");
        }
        if (!shouldPlay && mMediaPlayer.isPlaying()) {
            pause();
            Log.d(LOG_TAG, "setPlay: PAUSE");
        }
    }

    public void setMediaVolume(float volume) {
        if (volume < 0) {
            return;
        }
        if (mMediaPlayer == null) {
            return;
        }
        setVolume(volume, volume);
    }

    public void setCurrentTime(float currentTime) {
        float seekTime = currentTime * 1000;
        if (mMediaPlayer == null) {
            Log.d(LOG_TAG, "MEDIA PLAYER IS NULL");
            return;
        }
        int duration = getDuration();
        if (currentTime > duration || currentTime < 0) {
            seekTime = 0;
        }
        Log.d(LOG_TAG, "set seek to " + String.valueOf((int) seekTime));
        seekTo((int) seekTime);
    }

    public void setVideoEndAt(int endAt) {
        videoEndAt = endAt;
        if (mMediaPlayer == null) {
            return;
        }
        if (endAt > mMediaPlayer.getDuration() * 1000) {
            videoEndAt = mMediaPlayer.getDuration() * 1000;
        }
        if (mMediaPlayer.getCurrentPosition() * 1000 > videoEndAt) {
            mMediaPlayer.seekTo(videoStartAt);
        }
    }

    public void cleanup() {
        if (mMediaPlayer == null) {
            return;
        }
        mMediaPlayer.stop();
        mMediaPlayer.release();
        mMediaPlayer = null;
    }

    public void setVideoStartAt(int startAt) {
        videoStartAt = startAt;
        if (mMediaPlayer == null) {
            return;
        }
        if (mMediaPlayer.getDuration() * 1000 < videoStartAt) {
            mMediaPlayer.seekTo(videoStartAt);
        }
    }

    public void setProgressUpdateHandlerDelay(int delay) {
        this.progressUpdateHandlerDelay = delay;
    }

    public void sendMediaInfo() {
        if (mMediaPlayer == null) {
            Log.d(LOG_TAG, "sendMediaInfo: media Player is null");
            return;
        }
        int videoWidth = getVideoWidth();
        int videoHeight = getVideoHeight();

        WritableMap event = Arguments.createMap();

        event.putInt(Events.DURATION, mMediaPlayer.getDuration() / 1000);
        event.putInt(Events.WIDTH, videoWidth);
        event.putInt(Events.HEIGHT, videoHeight);

        eventEmitter.receiveEvent(getId(), EventsEnum.EVENT_GET_PREVIEW_IMAGE.toString(), event);
    }

    public void getFrame(float sec) {
        Bitmap bmp = metadataRetriever.getFrameAtTime((long) (sec * 1000000));
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        bmp.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream);
        byte[] byteArray = byteArrayOutputStream .toByteArray();
        String encoded = Base64.encodeToString(byteArray, Base64.DEFAULT);

        WritableMap event = Arguments.createMap();
        event.putString("image", encoded);

        eventEmitter.receiveEvent(getId(), EventsEnum.EVENT_GET_PREVIEW_IMAGE.toString(), event);
    }

    @Override
    public boolean onError(MediaPlayer mp, int what, int extra) {
        return false;
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);

        int videoWidth = getVideoWidth();
        int videoHeight = getVideoHeight();

        if (videoWidth == 0 || videoHeight == 0) {
            return;
        }
        Size viewSize = new Size(getWidth(), getHeight());
        Size videoSize = new Size(videoWidth, videoHeight);
        ScaleManager scaleManager = new ScaleManager(viewSize, videoSize);

        Matrix matrix = scaleManager.getScaleMatrix(mScalableType);
        if (matrix != null) {
            Log.d(LOG_TAG, "set transform");
            setTransform(matrix);
        }
    }

    @Override
    public void onPrepared(MediaPlayer mp) {
        videoEndAt = mp.getDuration() * 1000;
        Log.d(LOG_TAG, "onPrepared: " + videoEndAt);
    }

    @Override
    public void onBufferingUpdate(MediaPlayer mp, int percent) {

    }

    @Override
    public void onCompletion(MediaPlayer mp) {

    }

    @Override
    public boolean onInfo(MediaPlayer mp, int what, int extra) {
        return false;
    }

    @Override
    public void onHostResume() {

    }

    @Override
    public void onHostPause() {

    }

    @Override
    public void onHostDestroy() {

    }

    @Override
    public int getBufferPercentage() {
        return 0;
    }

    @Override
    public boolean canPause() {
        return false;
    }

    @Override
    public boolean canSeekBackward() {
        return false;
    }

    @Override
    public boolean canSeekForward() {
        return false;
    }

    @Override
    public int getAudioSessionId() {
        return 0;
    }
}
