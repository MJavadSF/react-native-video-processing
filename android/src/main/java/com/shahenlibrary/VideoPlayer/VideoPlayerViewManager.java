package com.shahenlibrary.VideoPlayer;

import android.util.Log;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.IllegalViewOperationException;
import com.facebook.react.uimanager.annotations.ReactProp;
import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.ThemedReactContext;
import com.shahenlibrary.Events.EventsEnum;
import com.shahenlibrary.Events.Events;

import java.util.Map;

import javax.annotation.Nullable;

public class VideoPlayerViewManager extends SimpleViewManager<VideoPlayerView> {
    private static final String REACT_PACKAGE = "RNVideoProcessing";
    private final String SET_SOURCE = "source";
    private final String SET_PLAY = "play";
    private final String SET_REPLAY = "replay";
    private final String SET_VOLUME = "volume";
    private final String SET_CURRENT_TIME = "currentTime";
    private final String SET_PROGRESS_DELAY = "progressEventDelay";
    private final String SET_VIDEO_END_TIME = "endTime";
    private final String SET_VIDEO_START_TIME = "startTime";

    private final int COMMAND_GET_INFO = 1;
    private final int COMMAND_TRIM_MEDIA = 2;
    private final int COMMAND_COMPRESS_MEDIA = 3;

    @Override
    public String getName() {
        return VideoPlayerViewManager.REACT_PACKAGE;
    }

    @Override
    protected VideoPlayerView createViewInstance(ThemedReactContext reactContext) {
        return new VideoPlayerView(reactContext);
    }

    @Override
    public void onDropViewInstance(VideoPlayerView player) {
        super.onDropViewInstance(player);
        player.cleanup();
    }

    @Nullable
    @Override
    public Map getExportedCustomDirectEventTypeConstants() {
        MapBuilder.Builder<String, Map> builder = MapBuilder.builder();
        for (EventsEnum evt : EventsEnum.values()) {
            builder.put(evt.toString(), MapBuilder.of("registrationName", evt.toString()));
        }
        Log.d(VideoPlayerViewManager.REACT_PACKAGE, builder.toString());
        return builder.build();
    }

    @Nullable
    @Override
    public Map<String, Integer> getCommandsMap() {
        Log.d(VideoPlayerViewManager.REACT_PACKAGE, "getCommandsMap");
        return MapBuilder.of(
                Events.COMPRESS_MEDIA,
                COMMAND_COMPRESS_MEDIA,

                Events.GET_MEDIA_INFO,
                COMMAND_GET_INFO,

                Events.TRIM_MEDIA,
                COMMAND_TRIM_MEDIA
        );
    }

    @Override
    public void receiveCommand(VideoPlayerView root, int commandId, @Nullable ReadableArray args) {
        switch (commandId) {
            case COMMAND_GET_INFO:
                root.sendMediaInfo();
                break;
            default:
                Log.d(VideoPlayerViewManager.REACT_PACKAGE, "receiveCommand: Wrong command received");
        }
    }

    @ReactProp(name = SET_SOURCE)
    public void setSource(final VideoPlayerView player, String source) {
        player.setSource(source);
    }

    @ReactProp(name = SET_PLAY, defaultBoolean = true)
    public void setPlay(final VideoPlayerView player, boolean shouldPlay) {
        Log.d(VideoPlayerViewManager.REACT_PACKAGE, "setPlay: " + String.valueOf(shouldPlay));
        player.setPlay(shouldPlay);
    }

    @ReactProp(name = SET_REPLAY, defaultBoolean = true)
    public void setReplay(final VideoPlayerView player, boolean replay) {
        Log.d(VideoPlayerViewManager.REACT_PACKAGE, "set replay: " + String.valueOf(replay));
        player.setLooping(replay);
    }

    @ReactProp(name = SET_VOLUME, defaultFloat = 10f)
    public void setVolume(final VideoPlayerView player, float volume) {
        Log.d(VideoPlayerViewManager.REACT_PACKAGE, "set volume: " + String.valueOf(volume));
        player.setMediaVolume(volume);
    }

    @ReactProp(name = SET_CURRENT_TIME, defaultFloat = 0f)
    public void setCurrentTime(final VideoPlayerView player, float seekTime) {
        Log.d(VideoPlayerViewManager.REACT_PACKAGE, "set current time: " + String.valueOf(seekTime) );
        player.setCurrentTime(seekTime);
    }

    @ReactProp(name = SET_PROGRESS_DELAY, defaultInt = 1000)
    public void setProgressDelay(final VideoPlayerView player, int delay) {
        player.setProgressUpdateHandlerDelay(delay);
    }

    @ReactProp(name = SET_VIDEO_END_TIME)
    public void setVideoEndTime(final VideoPlayerView player, int endTime) {
        Log.d(VideoPlayerViewManager.REACT_PACKAGE, "setVideoEndTime: " + String.valueOf(endTime));
        player.setVideoEndAt(endTime);
    }

    @ReactProp(name = SET_VIDEO_START_TIME)
    public void setVideoStartTime(final VideoPlayerView player, int startTime) {
        Log.d(VideoPlayerViewManager.REACT_PACKAGE, "setVideoStartTime: " + String.valueOf(startTime));
        player.setVideoStartAt(startTime);
    }

    @ReactMethod
    public void getMediaInfo(Promise promise) {
        promise.resolve(10);
    }
}
