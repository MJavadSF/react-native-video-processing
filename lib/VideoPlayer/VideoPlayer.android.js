import React, { PureComponent, PropTypes } from 'react';
import {
  AppRegistry,
  StyleSheet,
  Text,
  View,
  requireNativeComponent,
  CameraRoll,
  UIManager,
  findNodeHandle
} from 'react-native';

export class VideoPlayer extends PureComponent {
  static propTypes = {
    ...View.propTypes,
    play: PropTypes.bool,
    replay: PropTypes.bool,
    volume: PropTypes.number,
    getVideoInfo: PropTypes.func,
    onVideoProgress: PropTypes.func,
    currentTime: PropTypes.number,
    endTime: PropTypes.number,
    startTime: PropTypes.number,
    progressEventDelay: PropTypes.number,
    source: PropTypes.string.isRequired
  };
  constructor(props) {
    super(props);

    this._receiveVideoInfo = this._receiveVideoInfo.bind(this);
    this.getInfoPromisesResolves = [];
  }
  getVideoInfo() {
    UIManager.dispatchViewManagerCommand(
        findNodeHandle(this),
        UIManager.RNVideoProcessing.Commands.getInfo,
        [],
    );
    return new Promise ((resolve) => {
      this.getInfoPromisesResolves.push(resolve);
    });
  }

  _receiveVideoInfo({ nativeEvent }) {
    this.getInfoPromisesResolves.forEach((resolve) => resolve(nativeEvent));
    this.getInfoPromisesResolves = [];
  }

  render() {
    const {
      source,
      play,
      onVideoProgress,
      currentTime,
      endTime,
      startTime,
      replay,
      volume,
      style
    } = this.props;
    return (
      <RNVideoPlayer
        style={style}
        source={source}
        play={play}
        onVideoProgress={onVideoProgress}
        getVideoInfo={this._receiveVideoInfo}
        currentTime={currentTime}
        endTime={endTime}
        startTime={startTime}
        replay={replay}
        volume={volume}
      />
    );
  }
}

const RNVideoPlayer = requireNativeComponent('RNVideoProcessing', VideoPlayer);
