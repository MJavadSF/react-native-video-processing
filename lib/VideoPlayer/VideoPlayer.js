import React, { PropTypes, Component } from 'react';
import { View, requireNativeComponent, NativeModules } from 'react-native';
import { getActualSource } from '../utils';
const PLAYER_COMPONENT_NAME = 'RNVideoProcessing';

const { RNVideoTrimmer } = NativeModules;

export class VideoPlayer extends Component {
	static propTypes = {
		source: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
		play: PropTypes.bool,
		replay: PropTypes.bool,
		rotate: PropTypes.bool,
		currentTime: PropTypes.number,
		volume: PropTypes.number,
		startTime: PropTypes.number,
		background_Color: PropTypes.string,
		endTime: PropTypes.number,
		playerWidth: PropTypes.number,
		playerHeight: PropTypes.number,
		onChange: PropTypes.func,
		...View.propTypes
	};
	static defaultProps = {
		play: false,
		replay: false,
		rotate: false,
		volume: 0.0,
		currentTime: 0,
		startTime: 0,
	};
	constructor(...args) {
		super(...args);
		this.state = {};
		this.trim = this.trim.bind(this);
		this.getPreviewForSecond = this.getPreviewForSecond.bind(this);
		this.getVideoInfo = this.getVideoInfo.bind(this);
		this._onChange = this._onChange.bind(this);
	}

	getPreviewForSecond(source, forSecond = 0) {
		const actualSource = getActualSource(source);
		return new Promise((resolve, reject) => {
			RNVideoTrimmer.getPreviewImageAtPosition(actualSource, forSecond, (err, base64) => {
				if (err) {
					return reject(err);
				}
				return resolve(base64);
			});
		});
	}

	getVideoInfo(source) {
		const actualSource = getActualSource(source);
		return new Promise((resolve, reject) => {
			RNVideoTrimmer.getAssetInfo(actualSource, (err, info) => {
				if (err) {
					return reject(err);
				}
				return resolve(info);
			});
		});
	}

	_onChange(event) {
		if (!this.props.onChange) {
			return;
		}
		this.props.onChange(event);
	}

	trim(source, start, end) {
		const actualSource = getActualSource(source);
		return new Promise((resolve, reject) => {
			RNVideoTrimmer.trim(actualSource, start, end, (err, output) => {
				if (err) {
					return reject(err);
				}
				return resolve(output);
			});
		});
	}

	render() {
		const {
			source,
			play,
			replay,
			rotate,
			currentTime,
			backgroundColor,
			startTime,
			endTime,
			playerWidth,
			playerHeight,
			volume,
			...viewProps
		} = this.props;

		const actualSource = getActualSource(source);
		return (
			<RNVideoPlayer
				source={actualSource}
				play={play}
				replay={replay}
				rotate={rotate}
				volume={volume}
				playerWidth={playerWidth}
				playerHeight={playerHeight}
				currentTime={currentTime}
				background_Color={backgroundColor}
				startTime={startTime}
				endTime={endTime}
				onChange={this._onChange}
				{...viewProps}
			/>
		);
	}
}

const RNVideoPlayer = requireNativeComponent(PLAYER_COMPONENT_NAME, VideoPlayer);
