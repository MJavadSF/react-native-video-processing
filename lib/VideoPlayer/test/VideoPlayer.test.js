/* global expect */

import React from 'react';
import { VideoPlayer } from '../VideoPlayer';
import renderer from 'react-test-renderer';

describe('VideoPlayer', () => {
	it('Should be defined', () => {
		expect(VideoPlayer).toBeDefined();
	});
	it('Should render correctly', () => {
		const mockRender = renderer.create(
			<VideoPlayer source={'somePath'} />
		);

		const jsonRenderedComponent = mockRender.toJSON();

		expect(jsonRenderedComponent).toMatchSnapshot();
	});
});
