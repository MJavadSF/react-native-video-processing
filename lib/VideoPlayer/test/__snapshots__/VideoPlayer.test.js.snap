exports[`VideoPlayer Should render correctly 1`] = `<View />`;
